### PR TITLE
Add fsck/repair ability to delete corrupted objects

### PR DESCRIPTION
**Pull request by aclamk**
Each object's metadata is stored in RocksDB as 1 or more keys.
If some keys are missing there fsck/repair used to leave it alone.
If repair finds an object that is missing any key then it deletes all keys that belonged to this object. 
It relies on other parts of fsck/repair to clean allocations/shared blobs/omaps.

This problem was discovered when recovering customer's corrupted database.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
jenkins test make check

**Review by ifed01 — COMMENTED** [review]
I'm not sure the proposed approach - delete the rest onode's pieces - is the best option to deal with such errors. IMO less destructive approach would be to recover missed shard and "fill" it with zeroes or dummy extent(s). 
This would provide less data loss since onode metadata and content in existing shards are preserved.

File: src/test/objectstore/store_test.cc [hunk 1]
```diff
@@ -9159,6 +9159,122 @@ TEST_P(StoreTestSpecificAUSize, BluestoreRepairTest) {
...
+      }
+    }
+    ASSERT_EQ(db->submit_transaction_sync(t), 0);
+    bstore->close_db_environment();
+    cerr << "objects damaged" << std::endl;
+
+    cerr << "fscking/fixing" << std::endl;
+    ASSERT_GE(bstore->fsck(false), 3);
+    ASSERT_LE(bstore->repair(false), 0);
+    ASSERT_EQ(bstore->fsck(false), 1); //still statfs wrong
```
Comment (ifed01) [inline src/test/objectstore/store_test.cc:9269]: IMO that's a imperfect  user experience to have a need for multiple fsck/repair runs to get a clean state...

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -8095,7 +8102,27 @@ BlueStore::OnodeRef BlueStore::fsck_check_objects_shallow(
 
   num_spanning_blobs += o->extent_map.spanning_blob_map.size();
 
-  o->extent_map.fault_range(db, 0, OBJECT_MAX_SIZE);
+  bool fully_loaded = o->extent_map.fault_range(db, 0, OBJECT_MAX_SIZE);
+  if (!fully_loaded) {
+    // object is corrupted
+    derr << "fsck error: " << oid << " failed to load some shards" << dendl;
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:8108]: I'd prefer log ouput prefixed with "fsck error:" to be in-sync with reported error count. I.e. amount of such lines in the log to be equal to the number of detected error which fsck reports on completion. Ths line violates that condition

**Comment by aclamk** [comment]
Dropped this PR in favor of #45535

**Pull request closed without merging**